File: src/domain/breakdown.ts
import { createToyBaseObject, type BaseObject, type IsoDateTime, type StableId } from "./common";
import type { ContextTaskType } from "./context-pack";

export type ConcernKind =
  | "outline"
  | "character"
  | "scene"
  | "setting"
  | "worldbuilding"
  | "pov"
  | "reader-contract"
  | "style"
  | "craft"
  | "relationship"
  | "foreshadowing"
  | "revision"
  | "custom";

export interface CoverageClaim {
  scope: string;
  unavailableScope: string[];
  evidenceRefIds: StableId[];
  confidence: "low" | "medium" | "high";
  rationale: string;
}

export interface FieldEvidenceRef {
  fieldPath: string;
  evidenceRefIds: StableId[];
  required: boolean;
}

export interface TransferRule {
  id: StableId;
  sourceFact: string;
  transferableMechanism: string;
  targetVariableSlots: string[];
  prohibitedResidue: string[];
  misuseRisks: string[];
}

export interface EvidenceCard extends BaseObject {
  type: "evidence-card";
  breakdownProjectId: StableId;
  breakdownScopeId: StableId;
  sourceLocator: string;
  observationKind: "observation" | "inference" | "assumption";
  observedTechnique: string;
  functionInSource: string;
  transferableMechanism: string;
  prohibitedCopying: string[];
  coverage: CoverageClaim;
  confidence: "low" | "medium" | "high";
  fieldEvidenceRefs: FieldEvidenceRef[];
}

export interface ReferenceMechanism extends BaseObject {
  type: "reference-mechanism";
  sourceEvidenceRefs: StableId[];
  mechanism: string;
  transferRuleIds: StableId[];
  variableSlots: string[];
  prohibitedResidue: string[];
  misuseRisks: string[];
  applicableTaskTypes: ContextTaskType[];
  applicableScope: string;
  currentWorkConflictChecks: string[];
}

export interface BreakdownConcern {
  id: StableId;
  kind: ConcernKind;
  label: string;
  requiredForRecipe: boolean;
  coverage: CoverageClaim;
  fieldEvidenceRefs: FieldEvidenceRef[];
  transferRuleIds: StableId[];
}

export interface BookSpecificModuleField {
  name: string;
  valueKind: "text" | "number" | "boolean" | "string-list" | "record-list";
  evidenceRequired: boolean;
  displayHint: "card" | "table" | "timeline" | "graph" | "hidden";
  migrationStrategy: "preserve" | "drop-if-unknown" | "manual-review";
}

export interface BookSpecificModule {
  id: StableId;
  label: string;
  appliesToConcernIds: StableId[];
  fields: BookSpecificModuleField[];
}

export interface ViewProjection {
  id: StableId;
  label: string;
  targetConcernId: StableId;
  sourceFieldPaths: string[];
  displayKind: "card" | "table" | "timeline" | "graph";
}

export interface CompletionGateCheck {
  id: StableId;
  targetRef: StableId;
  gateLevel: "setup" | "stage" | "volume" | "book";
  checkKind: string;
  blocking: boolean;
  passed: boolean;
  evidenceRefIds: StableId[];
  notes: string;
}

export interface BreakdownRecipe extends BaseObject {
  type: "breakdown-recipe";
  concerns: BreakdownConcern[];
  transferRules: TransferRule[];
  modules: BookSpecificModule[];
  viewProjections: ViewProjection[];
  completionGateChecks: CompletionGateCheck[];
}

export interface BreakdownProject extends BaseObject {
  type: "breakdown-project";
  sourceId: StableId;
  recipeId?: StableId;
  manualMethod?: string;
  objective: string;
  scopeIds: StableId[];
  outputRefIds: StableId[];
}

export interface BreakdownScope extends BaseObject {
  type: "breakdown-scope";
  breakdownProjectId: StableId;
  scopeLabel: string;
  locator: string;
  coverageIntent: string;
  exclusions: string[];
}

export function createToyCoverageClaim(): CoverageClaim {
  return {
    scope: "toy chapter",
    unavailableScope: ["later chapters"],
    evidenceRefIds: ["evidence-toy-pressure"],
    confidence: "medium",
    rationale: "The toy fixture covers one invented scene only.",
  };
}

export function createToyBreakdownRecipe(timestamp: IsoDateTime): BreakdownRecipe {
  return {
    ...createToyBaseObject("breakdown-recipe", "recipe-toy-longform", "Toy Longform Recipe", timestamp, "active"),
    concerns: [
      {
        id: "concern-character-toy",
        kind: "character",
        label: "Character profile",
        requiredForRecipe: true,
        coverage: createToyCoverageClaim(),
        fieldEvidenceRefs: [
          {
            fieldPath: "stableProfile.coreDrive",
            evidenceRefIds: ["evidence-toy-pressure"],
            required: true,
          },
        ],
        transferRuleIds: ["transfer-rule-toy"],
      },
    ],
    transferRules: [
      {
        id: "transfer-rule-toy",
        sourceFact: "Invented public pressure fact.",
        transferableMechanism: "A visible constraint reveals priority.",
        targetVariableSlots: ["constraint", "choice", "cost"],
        prohibitedResidue: ["No copied names", "No copied event sequence"],
        misuseRisks: ["Constraint appears after the choice."],
      },
    ],
    modules: [
      {
        id: "module-character-profile-v1",
        label: "Character profile v1",
        appliesToConcernIds: ["concern-character-toy"],
        fields: [
          {
            name: "coreDrive",
            valueKind: "text",
            evidenceRequired: true,
            displayHint: "card",
            migrationStrategy: "manual-review",
          },
        ],
      },
    ],
    viewProjections: [
      {
        id: "view-character-card-toy",
        label: "Character card",
        targetConcernId: "concern-character-toy",
        sourceFieldPaths: ["stableProfile.coreDrive"],
        displayKind: "card",
      },
    ],
    completionGateChecks: [
      {
        id: "gate-character-profile-toy",
        targetRef: "concern-character-toy",
        gateLevel: "setup",
        checkKind: "character-profile-evidence",
        blocking: true,
        passed: true,
        evidenceRefIds: ["evidence-toy-pressure"],
        notes: "Toy character concern has one evidence reference.",
      },
    ],
  };
}

export function createToyEvidenceCard(timestamp: IsoDateTime): EvidenceCard {
  return {
    ...createToyBaseObject(
      "evidence-card",
      "evidence-toy-pressure",
      "Toy pressure evidence",
      timestamp,
      "active",
    ),
    breakdownProjectId: "breakdown-project-toy",
    breakdownScopeId: "breakdown-scope-toy",
    sourceLocator: "chapter-1-scene-1",
    observationKind: "observation",
    observedTechnique: "A visible resource limit turns route choice into pressure.",
    functionInSource: "Creates pressure without copying events or names.",
    transferableMechanism: "Expose a scarce resource before the character commits.",
    prohibitedCopying: ["No names", "No dialogue", "No event sequence"],
    coverage: createToyCoverageClaim(),
    confidence: "medium",
    fieldEvidenceRefs: [
      {
        fieldPath: "observedTechnique",
        evidenceRefIds: ["evidence-toy-pressure"],
        required: true,
      },
    ],
  };
}

export function createToyReferenceMechanism(timestamp: IsoDateTime): ReferenceMechanism {
  return {
    ...createToyBaseObject(
      "reference-mechanism",
      "reference-mechanism-toy",
      "Toy scarcity before commitment",
      timestamp,
      "active",
    ),
    sourceEvidenceRefs: ["evidence-toy-pressure"],
    mechanism: "Scarcity before commitment",
    transferRuleIds: ["transfer-rule-toy"],
    variableSlots: ["constraint", "choice", "cost"],
    prohibitedResidue: ["No copied names", "No copied event sequence"],
    misuseRisks: ["Constraint appears after the choice."],
    applicableTaskTypes: ["scene-planning", "draft-generation"],
    applicableScope: "original scenes that need costed choices",
    currentWorkConflictChecks: [
      "Do not reuse if the current scene already resolved the cost.",
    ],
  };
}

File: crates/core/src/model/breakdown.rs
use serde::{Deserialize, Serialize};

use super::common::{BaseObject, StableId};
use super::context_pack::ContextTaskType;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ConcernKind {
    Outline,
    Character,
    Scene,
    Setting,
    Worldbuilding,
    Pov,
    ReaderContract,
    Style,
    Craft,
    Relationship,
    Foreshadowing,
    Revision,
    Custom,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ConfidenceLevel {
    Low,
    Medium,
    High,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ModuleFieldValueKind {
    Text,
    Number,
    Boolean,
    StringList,
    RecordList,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ModuleDisplayHint {
    Card,
    Table,
    Timeline,
    Graph,
    Hidden,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ViewProjectionDisplayKind {
    Card,
    Table,
    Timeline,
    Graph,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ModuleMigrationStrategy {
    Preserve,
    DropIfUnknown,
    ManualReview,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum CompletionGateLevel {
    Setup,
    Stage,
    Volume,
    Book,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ObservationKind {
    Observation,
    Inference,
    Assumption,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CoverageClaim {
    pub scope: String,
    pub unavailable_scope: Vec<String>,
    pub evidence_ref_ids: Vec<StableId>,
    pub confidence: ConfidenceLevel,
    pub rationale: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldEvidenceRef {
    pub field_path: String,
    pub evidence_ref_ids: Vec<StableId>,
    pub required: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferRule {
    pub id: StableId,
    pub source_fact: String,
    pub transferable_mechanism: String,
    pub target_variable_slots: Vec<String>,
    pub prohibited_residue: Vec<String>,
    pub misuse_risks: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceCard {
    #[serde(flatten)]
    pub base: BaseObject,
    pub breakdown_project_id: StableId,
    pub breakdown_scope_id: StableId,
    pub source_locator: String,
    pub observation_kind: ObservationKind,
    pub observed_technique: String,
    pub function_in_source: String,
    pub transferable_mechanism: String,
    pub prohibited_copying: Vec<String>,
    pub coverage: CoverageClaim,
    pub confidence: ConfidenceLevel,
    pub field_evidence_refs: Vec<FieldEvidenceRef>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferenceMechanism {
    #[serde(flatten)]
    pub base: BaseObject,
    pub source_evidence_refs: Vec<StableId>,
    pub mechanism: String,
    pub transfer_rule_ids: Vec<StableId>,
    pub variable_slots: Vec<String>,
    pub prohibited_residue: Vec<String>,
    pub misuse_risks: Vec<String>,
    pub applicable_task_types: Vec<ContextTaskType>,
    pub applicable_scope: String,
    pub current_work_conflict_checks: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BreakdownConcern {
    pub id: StableId,
    pub kind: ConcernKind,
    pub label: String,
    pub required_for_recipe: bool,
    pub coverage: CoverageClaim,
    pub field_evidence_refs: Vec<FieldEvidenceRef>,
    pub transfer_rule_ids: Vec<StableId>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BookSpecificModuleField {
    pub name: String,
    pub value_kind: ModuleFieldValueKind,
    pub evidence_required: bool,
    pub display_hint: ModuleDisplayHint,
    pub migration_strategy: ModuleMigrationStrategy,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BookSpecificModule {
    pub id: StableId,
    pub label: String,
    pub applies_to_concern_ids: Vec<StableId>,
    pub fields: Vec<BookSpecificModuleField>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ViewProjection {
    pub id: StableId,
    pub label: String,
    pub target_concern_id: StableId,
    pub source_field_paths: Vec<String>,
    pub display_kind: ViewProjectionDisplayKind,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionGateCheck {
    pub id: StableId,
    pub target_ref: StableId,
    pub gate_level: CompletionGateLevel,
    pub check_kind: String,
    pub blocking: bool,
    pub passed: bool,
    pub evidence_ref_ids: Vec<StableId>,
    pub notes: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BreakdownRecipe {
    #[serde(flatten)]
    pub base: BaseObject,
    pub concerns: Vec<BreakdownConcern>,
    pub transfer_rules: Vec<TransferRule>,
    pub modules: Vec<BookSpecificModule>,
    pub view_projections: Vec<ViewProjection>,
    pub completion_gate_checks: Vec<CompletionGateCheck>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BreakdownProject {
    #[serde(flatten)]
    pub base: BaseObject,
    pub source_id: StableId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recipe_id: Option<StableId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub manual_method: Option<String>,
    pub objective: String,
    pub scope_ids: Vec<StableId>,
    pub output_ref_ids: Vec<StableId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BreakdownScope {
    #[serde(flatten)]
    pub base: BaseObject,
    pub breakdown_project_id: StableId,
    pub scope_label: String,
    pub locator: String,
    pub coverage_intent: String,
    pub exclusions: Vec<String>,
}
